src/audio/crossover/crossover_generic.c

```diff
@@ -0,0 +1,291 @@
+#include <sof/audio/eq_iir/iir.h>
+#include <sof/audio/component.h>
+#include <sof/audio/format.h>
+#include <sof/audio/crossover/crossover.h>
+
+/*
+ * \brief Splits x into two based on the coefficients set in the lp
+ *        and hp filters. The output of the lp is in y1, the output of
+ *        the hp is in y2.
+ *
+ * As a side effect, this function mutates the delay values of both
+ * filters.
+ */
+static void crossover_generic_lr4_split(struct iir_state_df2t *lp,
+					struct iir_state_df2t *hp,
+					int32_t x, int32_t *y1, int32_t *y2)
+{
+	*y1 = crossover_generic_process_lr4(x, lp);
+	*y2 = crossover_generic_process_lr4(x, hp);
+}
+
+/*
+ * \brief Splits input signal into two and merges it back to it's
+ *        original form.
+ *
+ * With 3-way crossovers, one output goes through only one LR4 filter,
+ * whereas the other two go through two LR4 filters. This causes the signals
+ * to be out of phase. We need to pass the signal through another set of LR4
+ * filters to align back the phase.
+ */
+static void crossover_generic_lr4_merge(struct iir_state_df2t *lp,
+					struct iir_state_df2t *hp,
+					int32_t x, int32_t *y)
+{
+	int32_t z1, z2;
+
+	z1 = crossover_generic_process_lr4(x, lp);
+	z2 = crossover_generic_process_lr4(x, hp);
+	*y = sat_int32(((int64_t)z1) + z2);
+}
+
+static void crossover_generic_split_2way(int32_t in,
+					 int32_t out[],
+					 struct crossover_state *state)
+{
+	crossover_generic_lr4_split(&state->lowpass[0], &state->highpass[0],
+				    in, &out[0], &out[1]);
+}
+
+static void crossover_generic_split_3way(int32_t in,
+					 int32_t out[],
+					 struct crossover_state *state)
+{
+	int32_t z1, z2;
+
+	crossover_generic_lr4_split(&state->lowpass[0], &state->highpass[0],
+				    in, &z1, &z2);
+	/* Realign the phase of z1 */
+	crossover_generic_lr4_merge(&state->lowpass[1], &state->highpass[1],
+				    z1, &out[0]);
+	crossover_generic_lr4_split(&state->lowpass[2], &state->highpass[2],
+				    z2, &out[1], &out[2]);
+}
+
+static void crossover_generic_split_4way(int32_t in,
+					 int32_t out[],
+					 struct crossover_state *state)
+{
+	int32_t z1, z2;
+
+	crossover_generic_lr4_split(&state->lowpass[1], &state->highpass[1],
+				    in, &z1, &z2);
+	crossover_generic_lr4_split(&state->lowpass[0], &state->highpass[0],
+				    z1, &out[0], &out[1]);
+	crossover_generic_lr4_split(&state->lowpass[2], &state->highpass[2],
+				    z2, &out[2], &out[3]);
+}
+
+#if CONFIG_FORMAT_S16LE
+static void crossover_s16_default_pass(const struct comp_dev *dev,
+				       const struct comp_buffer *source,
+				       struct comp_buffer *sinks[],
+				       int32_t num_sinks,
+				       uint32_t frames)
+{
+	const struct audio_stream *source_stream = &source->stream;
+	int16_t *x;
+	int32_t *y;
+	int i, j;
+	int n = source_stream->channels * frames;
+
+	for (i = 0; i < n; i++) {
+		x = audio_stream_read_frag_s16(source_stream, i);
+		for (j = 0; j < num_sinks; j++) {
+			if (!sinks[j])
+				continue;
```

Comment:
Can you do sinks on the outer loop, that way it's checked less frequently and will be a faster loop.

Reply:
@lgirdwood I am not sure what you are seeing here? This way we only need to read once and then write n times. If we flipped this it would be read each sample n times and write n times.

Reply:
I think @cujomalainey is right. If we do sinks on the outer loop, we will need to call audio_stream_read_frag_s16() to read x per sink.

Reply:
ok, so I guess you always want to increment x regardless of sink. Makes sense.